core/src/main/java/com/oath/oak/NativeAllocator/OakNativeMemoryAllocator.java

```diff
@@ -116,14 +141,23 @@ public ByteBuffer allocate(int size) {
     public void free(ByteBuffer bb) {
         allocated.addAndGet(-(bb.remaining()));
         if (stats != null) stats.release(bb);
-        freeList.add(new AbstractMap.SimpleImmutableEntry<>(freeCounter.getAndIncrement(), bb));
+        freeList.add(
+            new AbstractMap.SimpleImmutableEntry<>(
+                freeCounter.getAndIncrement(), new Slice(INVALID_BLOCK_ID,bb)));
     }
 
+  public void freeSlice(Slice slice) {
+    allocated.addAndGet(-(slice.getByteBuffer().remaining()));
+    if (stats != null) stats.release(slice.getByteBuffer());
+    freeList.add(new AbstractMap.SimpleImmutableEntry<>(freeCounter.getAndIncrement(), slice));
+  }
+
     // Releases all memory allocated for this Oak (should be used as part of the Oak destruction)
     // Not thread safe, should be a single thread call. (?)
     @Override
     public void close() {
-        for (Block b : blocks) {
+        if (!closed.compareAndSet(false,true)) return;
```

Comment:
This is currently necessary as Memory Manager still has two memory allocators: for keys and for values (pointing to the same OakNativeMemoryAllocator). So OakNativeMemoryAllocator is closed twice. Of course it can be resolved in other way, but this extra-cautious behavior looks good to me, as one day OakNativeMemoryAllocator may be stand alone.